Sources/MapboxCoreNavigation/BillingHandler.swift

```diff
@@ -0,0 +1,379 @@
+
+     - important: This variable is safe to use from any thread.
+     - parameter uuid: Session UUID which is provided in `BillingHandler.beginBillingSession(for:uuid:)`.
+     */
+    func sessionState(uuid: UUID) -> SessionState {
+        lock.lock(); defer {
+            lock.unlock()
+        }
+
+        guard let session = _sessions[uuid] else {
+            return .stopped
+        }
+
+        if session.isPaused {
+            return .paused
+        }
+        else {
+            return .running
+        }
+    }
+
+    /// The token to use for service requests like `Directions` etc. 
+    var serviceSkuToken: String {
+        let sessionType = lock { _sessionTypeForRequests() }
+
+        if let sessionType = sessionType {
+            return billingService.getSKUTokenIfValid(for: sessionType)
+        }
+        else {
+            return ""
+        }
+    }
+
+    private init(service: BillingService) {
+        self.billingService = service
+    }
+
+    /**
+     Starts a new billing session of the given `sessionType` identified by `uuid`.
+
+     The `uuid` that is used to create a billing session must be provided in the following methods to perform
+     relevant changes to the started billing session:
+     - `BillingHandler.stopBillingSession(with:)`
+     - `BillingHandler.pauseBillingSession(with:)`
+     - `BillingHandler.resumeBillingSession(with:)`
+
+     - Parameters:
+     - sessionType: The type of the billing session.
+     - uuid: The unique identifier of the billing session.
+     */
+    func beginBillingSession(for sessionType: SessionType, uuid: UUID) {
+        lock.lock()
+
+        if var existingSession = _sessions[uuid] {
+            existingSession.isPaused = false
+            _sessions[uuid] = existingSession
+        }
+        else {
+            let session = Session(type: sessionType, isPaused: false)
+            _sessions[uuid] = session
+        }
+        lock.unlock()
+
+        let triggerBillingServiceEvents = billingService.getSessionStatus(for: sessionType) != .running
+        if triggerBillingServiceEvents {
+            billingService.triggerBillingEvent(onError: { error in
+                print(error)
+            })
+            billingService.beginBillingSession(for: sessionType, onError: { [weak self] error in
+                self?.failedToBeginBillingSession(with: uuid, with: error)
+            })
+        }
+    }
+
+    /// Stops the billing session identified by the `uuid`.
+    func stopBillingSession(with uuid: UUID) {
+        lock.lock()
+        guard let session = _sessions[uuid] else {
+            lock.unlock(); return
+        }
+        _sessions[uuid] = nil
+
+        let triggerBillingServiceEvents = !_hasSession(with: session.type)
+            && billingService.getSessionStatus(for: session.type) != .stopped
+        lock.unlock()
+
+        if triggerBillingServiceEvents {
+            billingService.stopBillingSession(for: session.type)
```

Comment:
Looks like we should also check if there are not active sessions in `_sessions` with the same `session.type`.
User can instantiate 2 `PassiveLocationManager`'s and then deallocate 1. This will call `stopBillingSession(for: .freeDrive)` but the second location manager will remain alive. Even `sessionState(uuid:)` will report that such session is still alive and updates will continue.